File: plugins/sulis-execution/sdk/typescript/src/resources/change.ts
import { resultPayload } from '../helpers.js';
import type {
  AsyncSubprocessTransport,
  SubprocessTransport,
  TransportConfig,
} from '../transport.js';
import type {
  ChangeAdoptResult,
  ChangeFinishResult,
  ChangeListResult,
  ChangeStartResult,
  ChangeStatusResult,
} from '../types.js';

const BINARY = 'sulis-change';

function common(config: TransportConfig): Record<string, unknown> {
  return { repo_root: config.repoRoot };
}

export interface ChangeStartParams {
  slug: string;
  primitive?: string;
  base?: string;
}

export interface ChangeAdoptParams {
  slug: string;
  primitive?: string;
  base?: string;
  mode?: 'forward' | 'rewrite';
  remote_ref?: string;
  force?: boolean;
}

export interface ChangeFinishParams {
  slug: string;
  primitive?: string;
  base?: string;
  merge?: boolean;
  pr?: boolean;
  no_cleanup?: boolean;
}

export class ChangeResource {
  constructor(
    private readonly transport: SubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  start(params: ChangeStartParams): ChangeStartResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'start', {
        ...common(this.config),
        primitive: 'feat',
        base: 'dev',
        ...params,
      }),
    ) as unknown as ChangeStartResult;
  }

  adopt(params: ChangeAdoptParams): ChangeAdoptResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'adopt', {
        ...common(this.config),
        primitive: 'feat',
        base: 'dev',
        mode: 'forward',
        remote_ref: 'origin/dev',
        force: false,
        ...params,
      }),
    ) as unknown as ChangeAdoptResult;
  }

  finish(params: ChangeFinishParams): ChangeFinishResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'finish', {
        ...common(this.config),
        primitive: 'feat',
        base: 'dev',
        merge: false,
        pr: false,
        no_cleanup: false,
        ...params,
      }),
    ) as unknown as ChangeFinishResult;
  }

  list(params: { base?: string } = {}): ChangeListResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'list', {
        ...common(this.config),
        base: 'dev',
        ...params,
      }),
    ) as unknown as ChangeListResult;
  }

  status(params: {
    slug: string;
    primitive?: string;
    base?: string;
  }): ChangeStatusResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'status', {
        ...common(this.config),
        primitive: 'feat',
        base: 'dev',
        ...params,
      }),
    ) as unknown as ChangeStatusResult;
  }
}

export class AsyncChangeResource {
  constructor(
    private readonly transport: AsyncSubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  async start(params: ChangeStartParams): Promise<ChangeStartResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'start', {
        ...common(this.config),
        primitive: 'feat',
        base: 'dev',
        ...params,
      }),
    ) as unknown as ChangeStartResult;
  }

  async adopt(params: ChangeAdoptParams): Promise<ChangeAdoptResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'adopt', {
        ...common(this.config),
        primitive: 'feat',
        base: 'dev',
        mode: 'forward',
        remote_ref: 'origin/dev',
        force: false,
        ...params,
      }),
    ) as unknown as ChangeAdoptResult;
  }

  async finish(params: ChangeFinishParams): Promise<ChangeFinishResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'finish', {
        ...common(this.config),
        primitive: 'feat',
        base: 'dev',
        merge: false,
        pr: false,
        no_cleanup: false,
        ...params,
      }),
    ) as unknown as ChangeFinishResult;
  }

  async list(params: { base?: string } = {}): Promise<ChangeListResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'list', {
        ...common(this.config),
        base: 'dev',
        ...params,
      }),
    ) as unknown as ChangeListResult;
  }

  async status(params: {
    slug: string;
    primitive?: string;
    base?: string;
  }): Promise<ChangeStatusResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'status', {
        ...common(this.config),
        primitive: 'feat',
        base: 'dev',
        ...params,
      }),
    ) as unknown as ChangeStatusResult;
  }
}

File: plugins/sulis-execution/sdk/python/sulis_execution/resources/change.py
"""Change resource — wraps `sulis-change` subcommands.

Note: sulis-change is project-independent (the change branch convention
is repo-wide, not project-scoped). Methods here pass only --repo-root,
not --project.
"""
from __future__ import annotations

from typing import Optional

from sulis_execution._helpers import _kwargs_to_params, _result_payload
from sulis_execution.transport import (
    AsyncSubprocessTransport,
    SubprocessTransport,
    TransportConfig,
)
from sulis_execution.types import (
    ChangeAdoptResult,
    ChangeFinishResult,
    ChangeListResult,
    ChangeStartResult,
    ChangeStatusResult,
)

BINARY = "sulis-change"


def _common(config: TransportConfig) -> dict:
    return {"repo_root": str(config.repo_root)}


class ChangeResource:
    def __init__(
        self, transport: SubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    def start(
        self, *, slug: str, primitive: str = "feat", base: str = "dev"
    ) -> ChangeStartResult:
        params = {
            **_common(self._config), "slug": slug, "primitive": primitive, "base": base,
        }
        envelope = self._transport.invoke(BINARY, "start", params)
        return ChangeStartResult.model_validate(_result_payload(envelope))

    def adopt(
        self,
        *,
        slug: str,
        primitive: str = "feat",
        base: str = "dev",
        mode: str = "forward",
        remote_ref: str = "origin/dev",
        force: bool = False,
    ) -> ChangeAdoptResult:
        params = {
            **_common(self._config),
            "slug": slug, "primitive": primitive, "base": base,
            "mode": mode, "remote_ref": remote_ref, "force": force,
        }
        envelope = self._transport.invoke(BINARY, "adopt", params)
        return ChangeAdoptResult.model_validate(_result_payload(envelope))

    def finish(
        self,
        *,
        slug: str,
        primitive: str = "feat",
        base: str = "dev",
        merge: bool = False,
        pr: bool = False,
        no_cleanup: bool = False,
    ) -> ChangeFinishResult:
        params = {
            **_common(self._config),
            "slug": slug, "primitive": primitive, "base": base,
            "merge": merge, "pr": pr, "no_cleanup": no_cleanup,
        }
        envelope = self._transport.invoke(BINARY, "finish", params)
        return ChangeFinishResult.model_validate(_result_payload(envelope))

    def list(self, *, base: str = "dev") -> ChangeListResult:
        params = {**_common(self._config), "base": base}
        envelope = self._transport.invoke(BINARY, "list", params)
        return ChangeListResult.model_validate(_result_payload(envelope))

    def status(
        self, *, slug: str, primitive: str = "feat", base: str = "dev"
    ) -> ChangeStatusResult:
        params = {
            **_common(self._config),
            "slug": slug, "primitive": primitive, "base": base,
        }
        envelope = self._transport.invoke(BINARY, "status", params)
        return ChangeStatusResult.model_validate(_result_payload(envelope))


class AsyncChangeResource:
    def __init__(
        self, transport: AsyncSubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    async def start(
        self, *, slug: str, primitive: str = "feat", base: str = "dev"
    ) -> ChangeStartResult:
        params = {
            **_common(self._config), "slug": slug, "primitive": primitive, "base": base,
        }
        envelope = await self._transport.invoke(BINARY, "start", params)
        return ChangeStartResult.model_validate(_result_payload(envelope))

    async def adopt(
        self,
        *,
        slug: str,
        primitive: str = "feat",
        base: str = "dev",
        mode: str = "forward",
        remote_ref: str = "origin/dev",
        force: bool = False,
    ) -> ChangeAdoptResult:
        params = {
            **_common(self._config),
            "slug": slug, "primitive": primitive, "base": base,
            "mode": mode, "remote_ref": remote_ref, "force": force,
        }
        envelope = await self._transport.invoke(BINARY, "adopt", params)
        return ChangeAdoptResult.model_validate(_result_payload(envelope))

    async def finish(
        self,
        *,
        slug: str,
        primitive: str = "feat",
        base: str = "dev",
        merge: bool = False,
        pr: bool = False,
        no_cleanup: bool = False,
    ) -> ChangeFinishResult:
        params = {
            **_common(self._config),
            "slug": slug, "primitive": primitive, "base": base,
            "merge": merge, "pr": pr, "no_cleanup": no_cleanup,
        }
        envelope = await self._transport.invoke(BINARY, "finish", params)
        return ChangeFinishResult.model_validate(_result_payload(envelope))

    async def list(self, *, base: str = "dev") -> ChangeListResult:
        params = {**_common(self._config), "base": base}
        envelope = await self._transport.invoke(BINARY, "list", params)
        return ChangeListResult.model_validate(_result_payload(envelope))

    async def status(
        self, *, slug: str, primitive: str = "feat", base: str = "dev"
    ) -> ChangeStatusResult:
        params = {
            **_common(self._config),
            "slug": slug, "primitive": primitive, "base": base,
        }
        envelope = await self._transport.invoke(BINARY, "status", params)
        return ChangeStatusResult.model_validate(_result_payload(envelope))
